ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -126,26 +126,35 @@ ThreadSafeModule *getThreadAOTMod(void) {
   return &ThreadAOTMod;
 }
 
-std::vector<Value *> get_trace_inputs(Function *F, uintptr_t BBIdx) {
+std::vector<Value *> get_trace_inputs(Module *AOTMod, char **FuncNames,
+                                      size_t *BBs, size_t Len) {
   std::vector<Value *> Vec;
-  auto It = F->begin();
-  // Skip to the first block in the trace which contains the `start_tracing`
-  // call.
-  std::advance(It, BBIdx);
-  BasicBlock *BB = &*It;
-  for (auto I = BB->begin(); I != BB->end(); I++) {
-    if (isa<CallInst>(I)) {
-      CallInst *CI = cast<CallInst>(&*I);
-      if (CI->getCalledFunction()->getName() == YKTRACE_START) {
-        // Skip first argument to start_tracing.
-        for (auto Arg = CI->arg_begin() + 1; Arg != CI->arg_end(); Arg++) {
-          Vec.push_back(Arg->get());
+
+  // Search for the YKTRACE_START call and extract its arguments.
+  for (size_t Idx = 0; Idx < Len; Idx++) {
+    Function *F = AOTMod->getFunction(FuncNames[Idx]);
+    if (F == nullptr) {
+      // unmappable blocks can't contain YKTRACE_START.
```

Comment:
I'm not sure how we can end up with a trace where the first block does not contain `start_tracing`. I assumed that's always the first block in the trace. Do you have an example?

Reply:
We are searching the functions in the AOTMod that the trace touches here.

I'm working on a change which might mean we can revert this hunk back to it's simpler form.

Reply:
Okay, but why don't we just search for `start_tracing` in block 0 of the main function, where I think it's guaranteed to be?

Reply:
It was due to a) trace holes, and b) misunderstanding.

3d85f6e5357a reverts this function.

Reply:
:+1: